clippy_lints/src/once_cell_lazy.rs

```diff
@@ -0,0 +1,277 @@
+use rustc_errors::Applicability;
+use rustc_hir::def::{DefKind, Res};
+use rustc_hir::def_id::DefId;
+use rustc_hir::{self as hir, BodyId, Expr, ExprKind, Item, ItemKind};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_session::impl_lint_pass;
+use rustc_span::Span;
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Lints when a `once_cell::sync::Lazy` type is declared as static variables.
+    ///
+    /// ### Why restrict this?
+    /// Such useage have been superseded by the `std::sync::LazyLock` type,
+    /// replacing them would reduce dependency of `once_cell` crate.
+    ///
+    /// ### Example
+    /// ```rust
+    /// use once_cell_lazy::once_cell_lazy;
+    /// use once_cell::sync::Lazy;
+    ///
+    /// once_cell_lazy! {
+    ///     static ref FOO: String = "foo".to_uppercase();
+    /// }
+    /// static BAR: Lazy<String> = Lazy::new(|| "BAR".to_lowercase());
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// static FOO: std::sync::LazyLock<String> = std::sync::LazyLock::new(|| "foo".to_uppercase());
+    /// static BAR: std::sync::LazyLock<String> = std::sync::LazyLock::new(|| "BAR".to_lowercase());
+    /// ```
+    #[clippy::version = "1.81.0"]
+    pub ONCE_CELL_LAZY,
+    restriction,
+    "using `once_cell::sync::Lazy` type that could be replaced by `std::sync::LazyLock`"
+}
+
+/// A list containing functions with coresponding replacements in `LazyLock`.
+///
+/// Some functions could be replaced as well if we have replaced `Lazy` to `LazyLock`,
+/// therefore after suggesting replace the type, we need to make sure the function calls can be
+/// replaced, otherwise the suggestions cannot be applied thus the applicability should be
+/// `Unspecified` or `MaybeIncorret`.
+static FUNCTION_REPLACEMENTS: &[(&str, Option<&str>)] = &[
+    ("once_cell::sync::Lazy::force", Some("std::sync::LazyLock::force")),
+    ("once_cell::sync::Lazy::force_mut", None),
+    ("once_cell::sync::Lazy::get", None),
+    ("once_cell::sync::Lazy::get_mut", None),
+    ("once_cell::sync::Lazy::new", Some("std::sync::LazyLock::new")),
+    // Note that `Lazy::into_value` is not in the list,
+    // because we only check for `static`s in this lint, and `into_value` attempts to take ownership
+    // of the parameter, which means it would fail natively.
+    // But do keep in mind that the equivalant replacement is called `LazyLock::into_inner`
+    // if somehow we decided to expand this lint to catch "non-static"s.
+];
+
+pub struct OnceCellLazy {
+    msrv: Msrv,
+    sugg_map: FxIndexMap<DefId, Option<String>>,
+    lazy_type_defs: FxIndexMap<DefId, LazyInfo>,
+}
+
+impl OnceCellLazy {
+    #[must_use]
+    pub fn new(msrv: Msrv) -> Self {
+        Self {
+            msrv,
+            sugg_map: FxIndexMap::default(),
+            lazy_type_defs: FxIndexMap::default(),
+        }
+    }
+}
+
+impl_lint_pass!(OnceCellLazy => [ONCE_CELL_LAZY]);
+
+macro_rules! ensure_prerequisite {
```

Comment:
Probably should exclude `no_std` crates as well.

Reply:
yes, that's included